File: backend/models/project.py
from . import db

_CN_NUMS = ['', '一', '二', '三', '四', '五', '六', '七', '八', '九', '十']


class Project(db.Model):
    __tablename__ = "projects"

    id = db.Column(db.Integer, primary_key=True)
    number = db.Column(db.String(50), unique=True)
    name = db.Column(db.String(200), nullable=False)
    round = db.Column(db.Integer, default=1)
    amount = db.Column(db.Float)
    line = db.Column(db.String(2))
    method = db.Column(db.String(30), default="")
    demand_dept = db.Column(db.String(100), default="")
    manage_dept = db.Column(db.String(100), default="")
    agency_code = db.Column(db.String(10), default="")
    officer = db.Column(db.String(50), default="")
    content = db.Column(db.Text, default="")
    bid_time = db.Column(db.String(50), default="")
    status = db.Column(db.String(20), default="立项")
    supervisor = db.Column(db.String(100), default="")
    representative = db.Column(db.String(200), default="")
    created_by = db.Column(db.String(100), default="")
    created_at = db.Column(db.String(30), default="")
    updated_at = db.Column(db.String(30), default="")
    is_draft = db.Column(db.Integer, default=0)
    can_open = db.Column(db.String(10), default="")
    bid_note = db.Column(db.Text, default="")
    category = db.Column(db.String(10), default="")
    year = db.Column(db.String(10), default="")
    is_deleted = db.Column(db.Integer, default=0)
    deleted_at = db.Column(db.String(30), default="")
    # 模块5 采购文件编制：两步确认（采购需求确认 / 采购文件确认）
    demand_confirmed = db.Column(db.Integer, default=0)
    demand_confirmed_by = db.Column(db.String(100), default="")
    demand_confirmed_at = db.Column(db.String(30), default="")
    doc_confirmed = db.Column(db.Integer, default=0)
    doc_confirmed_by = db.Column(db.String(100), default="")
    doc_confirmed_at = db.Column(db.String(30), default="")

    def to_dict(self):
        r = self.round or 1
        if r <= 1:
            display_name = self.name
        else:
            cn = _CN_NUMS[r] if r < len(_CN_NUMS) else str(r)
            display_name = f"{self.name}（第{cn}次）"
        return {
            "id": self.id,
            "number": self.number or "",
            "name": self.name,
            "round": r,
            "display_name": display_name,
            "amount": self.amount,
            "line": self.line or "",
            "method": self.method or "",
            "demand_dept": self.demand_dept or "",
            "manage_dept": self.manage_dept or "",
            "agency_code": self.agency_code or "",
            "officer": self.officer or "",
            "content": self.content or "",
            "bid_time": self.bid_time or "",
            "status": self.status or "",
            "supervisor": self.supervisor or "",
            "representative": self.representative or "",
            "created_by": self.created_by or "",
            "created_at": self.created_at or "",
            "updated_at": self.updated_at or "",
            "is_draft": bool(self.is_draft),
            "can_open": self.can_open or "",
            "bid_note": self.bid_note or "",
            "category": self.category or "",
            "year": self.year or "",
            "is_deleted": bool(self.is_deleted),
            "deleted_at": self.deleted_at or "",
            "demand_confirmed": bool(self.demand_confirmed),
            "demand_confirmed_by": self.demand_confirmed_by or "",
            "demand_confirmed_at": self.demand_confirmed_at or "",
            "doc_confirmed": bool(self.doc_confirmed),
            "doc_confirmed_by": self.doc_confirmed_by or "",
            "doc_confirmed_at": self.doc_confirmed_at or "",
        }

File: frontend/src/services/project.ts
import api from './api'

/**
 * 当前处理人：项目此刻卡在谁手上（后端 services/pending_owner.py）。
 * null = 本环节无人待办（已办完/已归档/草稿）。
 */
export interface PendingEntry {
  event: string
  label: string          // 「待确认采购需求（5.1）」
  role: 'officer' | 'agency' | string
  role_label: string     // 经办人 / 代理机构
  owner: string          // 系统账号（可能为空）
  owner_name: string     // 姓名 / 代理机构名
  is_reject: boolean     // 是被驳回后的返工
  since: string          // 停留起始（本轮最后一次审批动作时间）
  waiting_days: number | null
}
export type Pending = (PendingEntry & {
  current_round: number
  others: PendingEntry[]  // 并行待处理事项（授权函/更正公告/合同/考核等）
}) | null

export interface Project {
  id: number
  number: string
  name: string
  round: number
  display_name: string
  amount: number | null
  line: string
  method: string
  demand_dept: string
  manage_dept: string
  agency_code: string
  agency_name?: string
  agency_legal_rep?: string
  agency_phone?: string
  agency_address?: string
  officer: string
  content: string
  bid_time: string
  status: string
  is_draft: boolean
  can_open: string
  bid_note: string
  supervisor: string
  representative: string
  created_at: string
  updated_at: string
  category: string
  year: string
  is_deleted: boolean
  deleted_at: string
  demand_confirmed: boolean
  demand_confirmed_by: string
  demand_confirmed_at: string
  doc_confirmed: boolean
  doc_confirmed_by: string
  doc_confirmed_at: string
  doc_agency_contact: string
  doc_agency_phone: string
  package_count?: number
  current_round?: number
  current_stage?: string        // demand_confirm|doc_confirm|announce|bid_open|result|round_failed|done
  pending_contract?: number     // 已中标未签约的包数
  contract_rdweb_submitted?: boolean  // 合同已推送 rd-web 审签
  agency_rdweb_submitted?: boolean    // 代理协议已推送 rd-web 审签
  pending?: Pending                   // 当前处理人
}

export interface Agency {
  id: number
  code: string
  name: string
  active: number
}

export interface ProjectMeta {
  agencies: Agency[]
  methods: string[]
  statuses: string[]
}

export const getProjects = (deleted = false) =>
  api.get<{ ok: boolean; data: Project[]; total: number }>(
    deleted ? '/projects?deleted=1' : '/projects'
  )

// 授权函专用：所有处于开标窗口(bid_open)的代理项目（采购部全部可见、不按经办人隔离）
export const getBidOpenProjects = () =>
  api.get<{ ok: boolean; data: Project[]; total: number }>('/projects/bid-open')

export const restoreProject = (id: number) =>
  api.post<{ ok: boolean; message: string }>(`/projects/${id}/restore`, {})

export const getProjectMeta = () =>
  api.get<ProjectMeta & { ok: boolean }>('/projects/meta')

export const getProject = (id: number) =>
  api.get<{ ok: boolean; data: Project; agencies: Agency[]; methods: string[]; statuses: string[] }>(`/projects/${id}`)

export const createProject = (data: Record<string, unknown>) =>
  api.post<{ ok: boolean; message: string; data: Project }>('/projects', data)

export const updateProject = (id: number, data: Record<string, unknown>) =>
  api.put<{ ok: boolean; message: string; data: Project }>(`/projects/${id}`, data)

export const deleteProject = (id: number) =>
  api.delete<{ ok: boolean; message: string }>(`/projects/${id}`)

// ── 项目进展图 ────────────────────────────────────────────────
export interface ProgressNode {
  key: string
  label: string
  done: boolean
  at: string
  by: string
  result_value?: string          // 开标节点：可开标/流标
  ann_status?: string            // 公告节点：草稿/待确认/已确认
  files?: { name: string; at: string; by: string }[]
  packages?: Array<Record<string, unknown>>
}
export interface ProgressRound {
  round_number: number
  status: string
  nodes: ProgressNode[]
}
export interface ProjectProgress {
  project: {
    id: number; name: string; number: string; method: string
    current_round: number; current_stage: string
    pending?: Pending
  }
  rounds: ProgressRound[]
}

export const getProjectProgress = (id: number) =>
  api.get<{ ok: boolean; data: ProjectProgress }>(`/projects/${id}/progress`)
